src/sinks/loki.rs

```diff
@@ -0,0 +1,446 @@
+    }
+}
+
+impl HttpSink for LokiConfig {
+    type Input = (Labels, (i64, String));
+    type Output = Vec<(Labels, (i64, String))>;
+
+    fn encode_event(&self, mut event: Event) -> Option<Self::Input> {
+        let mut labels = Vec::new();
+
+        for (key, template) in &self.labels {
+            if let Ok(value) = template.render_string(&event) {
+                labels.push((key.clone(), value));
+            }
+
+            // Check if we need to remove any labels that might
+            // have been extracted. Defaults to false.
+            if self.remove_labels.unwrap_or(false) {
+                if let Some(fields) = template.get_fields() {
+                    for field in fields {
+                        event.as_mut_log().remove(&field);
+                    }
+                }
+            }
+        }
+
+        let ts = if let Some(event::Value::Timestamp(ts)) = event.as_log().get(&event::TIMESTAMP) {
+            ts.timestamp_nanos()
+        } else {
+            chrono::Utc::now().timestamp_nanos()
+        };
+
+        if self.remove_timestamp.unwrap_or(true) {
+            event.as_mut_log().remove(&event::TIMESTAMP);
+        }
+
+        let event = match &self.encoding {
+            Encoding::Json => serde_json::to_string(&event.as_log().all_fields())
+                .expect("json encoding should never fail"),
+
+            Encoding::Text => event
+                .as_log()
+                .get(&event::MESSAGE)
+                .map(Value::to_string_lossy)
+                .unwrap_or_default(),
+        };
+
+        // If no labels are provided we set our own default
+        // `{agent="vector"}` label. This can happen if the only
+        // label is a templatable one but the event doesn't match.
+        if labels.is_empty() {
+            labels = vec![("agent".to_string(), "vector".to_string())]
+        }
+
+        Some((labels, (ts, event)))
+    }
+
+    fn build_request(&self, events: Self::Output) -> http::Request<Vec<u8>> {
+        let mut streams: HashMap<Labels, Vec<(i64, String)>> = HashMap::new();
+
+        for (mut labels, event) in events {
+            // We must sort here to ensure it hashes to the same stream
+            // if the label set matches.
+            labels.sort();
+
+            if let Some(stream) = streams.get_mut(&labels) {
+                stream.push(event);
+            } else {
+                streams.insert(labels, vec![event]);
+            }
+        }
+
+        // Construct the json body
+        let mut streams_json: Vec<serde_json::Value> = Vec::new();
+
+        for (stream, mut events) in streams {
+            // Sort by timestamp
+            events.sort_by(|e1, e2| e1.0.cmp(&e2.0));
+
+            let stream = stream.into_iter().collect::<HashMap<_, _>>();
+            let events = events
+                .into_iter()
+                // The final json output should be: `[ts, line]`
+                .map(|e| json!([format!("{}", e.0), e.1]))
```

Comment:
Would it be better to explicitly format the timestamp, e.g. `to_rfc3339()`?

Reply:
This is already an `i64` but Loki wants it as a string instead of a json number.